spec/child-apps/happy-unload/happy-unload.spec.js

```diff
@@ -0,0 +1,219 @@
+					expect(childApp.getNumBootstrapCalls()).toBe(1);
+					expect(childApp.getNumMountCalls()).toBe(1);
+					expect(childApp.getNumUnmountCalls()).toBe(1);
+					expect(childApp.getNumUnloadCalls()).toBe(0);
+					return singleSpa.unloadChildApplication('./happy-unload.app.js');
+				})
+				.then(() => {
+					expect(singleSpa.getAppStatus('./happy-unload.app.js')).toEqual('NOT_LOADED');
+					expect(childApp.getNumBootstrapCalls()).toBe(1);
+					expect(childApp.getNumMountCalls()).toBe(1);
+					expect(childApp.getNumUnmountCalls()).toBe(1);
+					expect(childApp.getNumUnloadCalls()).toBe(1);
+					done();
+				})
+				.catch(err => {
+					fail(err);
+					done();
+				});
+			});
+		});
+
+		describe(`when waitForUnmount = true :`, () => {
+			it(`delays unloading the app until the app is no longer active`, done => {
+				window.location.hash = activeHash;
+
+				singleSpa
+				.triggerAppChange()
+				.then(() => {
+					expect(singleSpa.getAppStatus('./happy-unload.app.js')).toEqual('MOUNTED');
+					expect(childApp.getNumBootstrapCalls()).toBe(1);
+					expect(childApp.getNumMountCalls()).toBe(1);
+					expect(childApp.getNumUnmountCalls()).toBe(0);
+					expect(childApp.getNumUnloadCalls()).toBe(0);
+
+					singleSpa
+					.unloadChildApplication('./happy-unload.app.js', {waitForUnmount: true})
+					.then(() => {
+						/* This will get called only once the app is unloaded. And it will not
+						 * wait for the app to get remounted before it is called.
+						 */
+						expect(singleSpa.getAppStatus('./happy-unload.app.js')).toEqual('NOT_LOADED');
+
+						expect(childApp.getNumBootstrapCalls()).toBe(1);
+						expect(childApp.getNumMountCalls()).toBe(1);
+						expect(childApp.getNumUnmountCalls()).toBe(1);
+						expect(childApp.getNumUnloadCalls()).toBe(1);
+					})
+					.catch(err => {
+						fail(err);
+						done();
+					})
+
+					/* Triggering an app change after calling unloadChildApplication will
+					 * not cause the app to unload, since it is still mounted and we set
+					 * waitForUnmount: true.
+					 */
+					return singleSpa.triggerAppChange();
+				})
+				.then(() => {
+					expect(singleSpa.getAppStatus('./happy-unload.app.js')).toEqual('MOUNTED');
+					expect(childApp.getNumBootstrapCalls()).toBe(1);
+					expect(childApp.getNumMountCalls()).toBe(1);
+					expect(childApp.getNumUnmountCalls()).toBe(0);
+					expect(childApp.getNumUnloadCalls()).toBe(0);
+
+					window.location.hash = `#`;
+					return singleSpa.triggerAppChange();
+				})
+				.then(() => {
+					expect(singleSpa.getAppStatus('./happy-unload.app.js')).toEqual('NOT_LOADED');
+					expect(childApp.getNumBootstrapCalls()).toBe(1);
+					expect(childApp.getNumMountCalls()).toBe(1);
+					expect(childApp.getNumUnmountCalls()).toBe(1);
+					expect(childApp.getNumUnloadCalls()).toBe(1);
+					done();
+				})
+				.catch(err => {
+					fail(err);
+					done();
+				});
+			});
+		});
+
```

Comment:
It looks like right now these test unloading when the app is in `MOUNTED` and `NOT_LOADED` status.  Would it be worthwhile to add a few more tests for calling `unloadChildApplication` when an app is in other states?
Specifically for each of the life-cycle status'?

`LOADING_SOURCE_CODE`
`NOT_BOOTSTRAPPED`
`BOOTSTRAPPING`
`NOT_MOUNTED`
`MOUNTING`
`UNMOUNTING`
`UNLOADING`